### ISSUE TITLE
Internal promise rejection when using an unauthorised but valid key

### ISSUE THREAD

**Issue by stuartlangridge** [51Degrees/pipeline-node#2]
If an API key is used that does not have permission to make the calls, the fiftyone.pipeline.core module throws an internal promise rejection.

Below a key is used which causes this to fail.

```javascript

const CloudRequestEngine = require('fiftyone.pipeline.cloudrequestengine');
const PipelineBuilder = require('fiftyone.pipeline.core').PipelineBuilder;
const engines = require('fiftyone.pipeline.engines');
const core = require('fiftyone.pipeline.core');
const Engine = engines.Engine;
const AspectDataDictionary = engines.AspectDataDictionary;
const AspectPropertyValue = core.AspectPropertyValue;

class SimpleCloud extends Engine {
  constructor () {
    super(...arguments);
    this.dataKey = 'location';
  }

  processInternal (flowData) {
    const engine = this;
    this.checkProperties(flowData).then(function (params) {
      let cloudData = flowData.get('cloud').get('cloud');
      cloudData = JSON.parse(cloudData);
      const result = {};
      Object.entries(cloudData[engine.dataKey]).forEach(function ([key, value]) {
        result[key] = new AspectPropertyValue();
        if (cloudData[engine.dataKey][key + 'nullreason']) {
          result[key].noValueMessage = cloudData[engine.dataKey][key + 'nullreason'];
        } else {
          result[key].value = value;
        }
      });
      const data = new AspectDataDictionary({flowElement: engine, contents: result});
      flowData.setElementData(data);
    });
  }

  checkProperties (flowData) {
    const engine = this;
    return new Promise(function (resolve, reject) {
      // Check if properties set, if not set them
      if (!Object.keys(engine.properties).length) {
        const cloudProperties = flowData.get('cloud').get('properties');
        const locationProperties = cloudProperties.location;
        engine.properties = locationProperties;
        engine.updateProperties().then(resolve);
      } else {
        resolve();
      }
    });
  }
}


class SimplePipelineBuilder extends PipelineBuilder {
  constructor ({resourceKey = null}) {
    super(...arguments);
    this.flowElements.push(new CloudRequestEngine({ resourceKey: resourceKey }));
    this.flowElements.push(new SimpleCloud());
  }
}

let localResourceKey = 'AQQNX4tozS1CfutY2Eg';
const pipeline = new SimplePipelineBuilder({resourceKey: localResourceKey}).build();

const getCountry = async function (latitude, longitude) {
  const flowData = pipeline.createFlowData();
  flowData.evidence.add('query.51D_Pos_latitude', latitude);
  flowData.evidence.add('query.51D_Pos_longitude', longitude);
  await flowData.process();
  console.log(flowData.location); // no errors are indicated
};

getCountry('51', '-1');
```

This seems to be because, with this unauthorised (but valid) key, no properties are available in pipeline.js.

A simple fix is:

```
--- node_modules/fiftyone.pipeline.core/pipeline.js  1985-10-26 09:15:00.000000000 +0100
+++ node_modules_fixed/fiftyone.pipeline.core/pipeline.js  2020-10-10 18:24:38.424294200 +0100
@@ -242,8 +243,7 @@
 
       Promise.resolve(flowElement.getProperties()).then(function (properties) {
         const flowElementDataKey = flowElement.dataKey;

-        Object.entries(properties).forEach(function ([
+        Object.entries(properties || {}).forEach(function ([
           propertyKey,
           propertyMeta
         ]) {
```

(that is, quietly handle properties being null or undefined, rather than assuming that it's an object). There may be a better way to fix this internally to the module (which is why this is not a proposed PR).

**Comment by tungntpham** [comment]
Hi @stuartlangridge . Many thanks. We will release the fix for this in the v4.2.0 version.

**Comment by tungntpham** [comment]
Version 4.2.0 is now released.
